src/mlpack/methods/kde/kde_impl.hpp

```diff
@@ -0,0 +1,474 @@
+{
+  if (trained)
+  {
+    if (ownsReferenceTree)
+      referenceTree = new Tree(other.referenceTree);
+    else
+      referenceTree = other.referenceTree;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+KDE<MetricType, MatType, KernelType, TreeType>::KDE(KDE&& other) :
+    kernel(other.kernel),
+    metric(other.metric),
+    referenceTree(other.referenceTree),
+    relError(other.relError),
+    absError(other.absError),
+    breadthFirst(other.breadthFirst),
+    ownsKernel(other.ownsKernel),
+    ownsMetric(other.ownsMetric),
+    ownsReferenceTree(other.ownsReferenceTree),
+    trained(other.trained)
+{
+  other.kernel = new KernelType();
+  other.metric = new MetricType();
+  other.referenceTree = nullptr;
+  other.ownsReferenceTree = false;
+  other.trained = false;
+}
+
+template<typename MetricType,
+         typename MatType,
+         typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+KDE<MetricType, MatType, KernelType, TreeType>&
+KDE<MetricType, MatType, KernelType, TreeType>::operator=(KDE other)
+{
+  // Clean memory
+  if (ownsKernel)
+    delete kernel;
+  if (ownsMetric)
+    delete metric;
+  if (ownsReferenceTree)
+    delete referenceTree;
+
+  // Move
+  this->kernel = std::move(other.kernel);
+  this->metric = std::move(other.metric);
+  this->referenceTree = std::move(other.referenceTree);
+  this->relError = other.relError;
+  this->absError = other.absError;
+  this->breadthFirst = other.breadthFirst;
+  this->ownsKernel = other.ownsKernel;
+  this->ownsMetric = other.ownsMetric;
+  this->ownsReferenceTree = other.ownsReferenceTree;
+  this->trained = other.trained;
+
+  return *this;
+}
+
+template<typename MetricType,
+         typename MatType,
+         typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+KDE<MetricType, MatType, KernelType, TreeType>::~KDE()
+{
+  if (ownsKernel)
+    delete kernel;
+  if (ownsMetric)
+    delete metric;
+  if (ownsReferenceTree)
+    delete referenceTree;
+}
+
+template<typename MetricType,
+         typename MatType,
+         typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+void KDE<MetricType, MatType, KernelType, TreeType>::
+Train(MatType referenceSet)
+{
+  // Check if referenceSet is not an empty set.
+  if (referenceSet.n_cols == 0)
+    throw std::invalid_argument("cannot train KDE model with an empty "
+                                "reference set");
+  this->ownsReferenceTree = true;
+  this->referenceTree = new Tree(referenceSet);
```

Comment:
I think this should be `new Tree(std::move(referenceSet))`; correct me if I'm wrong. :)